### ISSUE TITLE
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by [rank0]: making sure all `forward` function outputs participate in calculating loss.

### ISSUE THREAD

**Issue by AryanParekh** [ultralytics/ultralytics#14093]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Multi-GPU

### Bug

So, I am facing an issue with the RTDETR model when I'm training it on the custom dataset. I've already executed 50 epochs of YOLOv8x, YOLOv8l, YOLOv8m, YOLOv8s and YOLOv8n, so I'm sure there's nothing wrong with the way my dataset has been configured. I haven't made any internal changes to the structure of the model, and I'm also using the latest version of Ultralytics. The execution stops after 12 epochs of RTDETRx and 13 epochs of the RTDETRl model. I've trained similar datasets with similar bounding box annotations for 200 eopchs on all these models twice, but I haven't faced this issue before.

```
Ultralytics YOLOv8.2.45 🚀 Python-3.11.5 torch-2.3.0 CUDA:0 (Tesla V100-SXM2-32GB, 32494MiB)
                                                     CUDA:1 (Tesla V100-SXM2-32GB, 32494MiB)
engine/trainer: task=detect, mode=train, model=models/rtdetr/rtdetr-x.pt, data=datasets/ROADLAB_TrafficSignalClassification/roadlabTSC.yaml, epochs=50, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=[0, 1], workers=8, project=None, name=RTDETRx_RoadlabSignal_50, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs/detect/RTDETRx_RoadlabSignal_50
Overriding model.yaml nc=80 with nc=5
WARNING ⚠️ no model scale passed. Assuming scale='x'.

                   from  n    params  module                                       arguments
  0                  -1  1     25792  ultralytics.nn.modules.block.HGStem          [3, 32, 64]
  1                  -1  6    259200  ultralytics.nn.modules.block.HGBlock         [64, 64, 128, 3, 6]
  2                  -1  1      1408  ultralytics.nn.modules.conv.DWConv           [128, 128, 3, 2, 1, False]
  3                  -1  6   1248256  ultralytics.nn.modules.block.HGBlock         [128, 128, 512, 3, 6]
  4                  -1  6   1788928  ultralytics.nn.modules.block.HGBlock         [512, 128, 512, 3, 6, False, True]
  5                  -1  1      5632  ultralytics.nn.modules.conv.DWConv           [512, 512, 3, 2, 1, False]
  6                  -1  6   2079232  ultralytics.nn.modules.block.HGBlock         [512, 256, 1024, 5, 6, True, False]
  7                  -1  6   2472448  ultralytics.nn.modules.block.HGBlock         [1024, 256, 1024, 5, 6, True, True]
  8                  -1  6   2472448  ultralytics.nn.modules.block.HGBlock         [1024, 256, 1024, 5, 6, True, True]
  9                  -1  6   2472448  ultralytics.nn.modules.block.HGBlock         [1024, 256, 1024, 5, 6, True, True]
 10                  -1  6   2472448  ultralytics.nn.modules.block.HGBlock         [1024, 256, 1024, 5, 6, True, True]
 11                  -1  1     11264  ultralytics.nn.modules.conv.DWConv           [1024, 1024, 3, 2, 1, False]
 12                  -1  6   8221696  ultralytics.nn.modules.block.HGBlock         [1024, 512, 2048, 5, 6, True, False]
 13                  -1  6   9794560  ultralytics.nn.modules.block.HGBlock         [2048, 512, 2048, 5, 6, True, True]
 14                  -1  1    787200  ultralytics.nn.modules.conv.Conv             [2048, 384, 1, 1, None, 1, 1, False]
 15                  -1  1   2168192  ultralytics.nn.modules.transformer.AIFI      [384, 2048, 8]
 16                  -1  1    148224  ultralytics.nn.modules.conv.Conv             [384, 384, 1, 1]
 17                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']
 18                  10  1    393984  ultralytics.nn.modules.conv.Conv             [1024, 384, 1, 1, None, 1, 1, False]
 19            [-2, -1]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 20                  -1  3   5019648  ultralytics.nn.modules.block.RepC3           [768, 384, 3]
 21                  -1  1    148224  ultralytics.nn.modules.conv.Conv             [384, 384, 1, 1]
 22                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']
 23                   4  1    197376  ultralytics.nn.modules.conv.Conv             [512, 384, 1, 1, None, 1, 1, False]
 24            [-2, -1]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 25                  -1  3   5019648  ultralytics.nn.modules.block.RepC3           [768, 384, 3]
 26                  -1  1   1327872  ultralytics.nn.modules.conv.Conv             [384, 384, 3, 2]
 27            [-1, 21]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 28                  -1  3   5019648  ultralytics.nn.modules.block.RepC3           [768, 384, 3]
 29                  -1  1   1327872  ultralytics.nn.modules.conv.Conv             [384, 384, 3, 2]
 30            [-1, 16]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 31                  -1  3   5019648  ultralytics.nn.modules.block.RepC3           [768, 384, 3]
 32        [25, 28, 31]  1   7410431  ultralytics.nn.modules.head.RTDETRDecoder    [5, [384, 384, 384]]
rt-detr-x summary: 867 layers, 67313727 parameters, 67313727 gradients, 232.4 GFLOPs

Transferred 1226/1241 items from pretrained weights
DDP: debug command /localscratch/aparek8.35464805.0/ultralyticsENV/bin/python -m torch.distributed.run --nproc_per_node 2 --master_port 47929 /home/aparek8/.config/Ultralytics/DDP/_temp_jkws7orf22759235773008.py
Ultralytics YOLOv8.2.45 🚀 Python-3.11.5 torch-2.3.0 CUDA:0 (Tesla V100-SXM2-32GB, 32494MiB)
                                                     CUDA:1 (Tesla V100-SXM2-32GB, 32494MiB)
Overriding model.yaml nc=80 with nc=5
WARNING ⚠️ no model scale passed. Assuming scale='x'.
Transferred 1226/1241 items from pretrained weights
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
train: Scanning /project/6000603/aparek8/ultralytics_CustomData/datasets/ROADLAB_TrafficSignalClassification/labels/train.cache... 22979 images, 0 backgrounds, 0 corrupt: 100%|██████████| 22979/22979 [00:train: Scanning /project/6000603/aparek8/ultralytics_CustomData/datasets/ROADLAB_TrafficSignalClassification/labels/train.cache... 22979 images, 0 backgrounds, 0 corrupt: 100%|██████████| 22979/22979 [00:00<?, ?it/s]
val: Scanning /project/6000603/aparek8/ultralytics_CustomData/datasets/ROADLAB_TrafficSignalClassification/labels/val.cache... 5803 images, 0 backgrounds, 0 corrupt: 100%|██████████| 5803/5803 [00:00<?, ?val: Scanning /project/6000603/aparek8/ultralytics_CustomData/datasets/ROADLAB_TrafficSignalClassification/labels/val.cache... 5803 images, 0 backgrounds, 0 corrupt: 100%|██████████| 5803/5803 [00:00<?, ?it/s]
Plotting labels to runs/detect/RTDETRx_RoadlabSignal_50/labels.jpg...
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically...
optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 193 weight(decay=0.0), 256 weight(decay=0.0005), 276 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 16 dataloader workers
Logging results to runs/detect/RTDETRx_RoadlabSignal_50
Starting training for 50 epochs...

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
       1/50      9.95G      1.128      5.313     0.3069          5        640: 100%|██████████| 1437/1437 [11:46<00:00,  2.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:09<00:00,  5.26it/s]
                   all       5803       9174      0.328      0.178      0.114     0.0404

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
       2/50      10.3G     0.7654     0.8025     0.1604          5        640: 100%|██████████| 1437/1437 [11:07<00:00,  2.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:08<00:00,  5.30it/s]
                   all       5803       9174      0.494      0.288      0.202     0.0654

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
       3/50      10.5G     0.8022     0.7617     0.1623          3        640: 100%|██████████| 1437/1437 [11:07<00:00,  2.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:08<00:00,  5.31it/s]
                   all       5803       9174      0.289      0.331      0.207      0.072

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
       4/50      10.5G     0.7746     0.7965     0.1603          8        640: 100%|██████████| 1437/1437 [10:59<00:00,  2.18it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:08<00:00,  5.28it/s]
                   all       5803       9174      0.307      0.331      0.223      0.081

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
       5/50      10.2G     0.7389     0.7768     0.1508          2        640: 100%|██████████| 1437/1437 [11:10<00:00,  2.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:08<00:00,  5.27it/s]
                   all       5803       9174      0.375       0.34      0.244     0.0917

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
       6/50      10.4G      0.695     0.7731     0.1422          3        640: 100%|██████████| 1437/1437 [11:15<00:00,  2.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:08<00:00,  5.29it/s]
                   all       5803       9174        0.4      0.364      0.273      0.105

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
       7/50      10.5G     0.6886     0.7634     0.1406          3        640: 100%|██████████| 1437/1437 [10:55<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:07<00:00,  5.38it/s]
                   all       5803       9174      0.386      0.361      0.252      0.107

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
       8/50      10.5G     0.6828     0.7574     0.1382          3        640: 100%|██████████| 1437/1437 [11:09<00:00,  2.15it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:07<00:00,  5.35it/s]
                   all       5803       9174      0.386       0.39      0.271      0.107

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
       9/50      10.2G     0.6744     0.7519     0.1354          5        640: 100%|██████████| 1437/1437 [11:11<00:00,  2.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:14<00:00,  4.88it/s]
                   all       5803       9174        0.4      0.375      0.268      0.106

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
      10/50      10.5G     0.6613      0.746     0.1311          3        640: 100%|██████████| 1437/1437 [11:10<00:00,  2.14it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:08<00:00,  5.29it/s]
                   all       5803       9174      0.402      0.401      0.294      0.118

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
      11/50      10.4G     0.6517     0.7387     0.1306          9        640: 100%|██████████| 1437/1437 [10:57<00:00,  2.19it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:08<00:00,  5.29it/s]
                   all       5803       9174      0.399      0.402      0.283       0.11

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/autograd/graph.py:744: UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'. You can file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation. (Triggered internally at /home/coulombc/wheels_builder/tmp.2617/python-3.11/torch/aten/src/ATen/Context.cpp:83.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
      12/50      10.4G     0.6565     0.7291     0.1304          0        640: 100%|██████████| 1437/1437 [11:15<00:00,  2.13it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 363/363 [01:08<00:00,  5.30it/s]
                   all       5803       9174      0.424      0.414      0.299      0.114

      Epoch    GPU_mem  giou_loss   cls_loss    l1_loss  Instances       Size
  0%|          | 0/1437 [00:00<?, ?it/s]
[rank0]: Traceback (most recent call last):
[rank0]:   File "/home/aparek8/.config/Ultralytics/DDP/_temp_jkws7orf22759235773008.py", line 13, in <module>
[rank0]:     results = trainer.train()
[rank0]:               ^^^^^^^^^^^^^^^
[rank0]:   File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/ultralytics/engine/trainer.py", line 204, in train
[rank0]:     self._do_train(world_size)
[rank0]:   File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/ultralytics/engine/trainer.py", line 381, in _do_train
[rank0]:     self.loss, self.loss_items = self.model(batch)
[rank0]:                                  ^^^^^^^^^^^^^^^^^
[rank0]:   File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/nn/parallel/distributed.py", line 1589, in forward
[rank0]:     inputs, kwargs = self._pre_forward(*inputs, **kwargs)
[rank0]:                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/nn/parallel/distributed.py", line 1480, in _pre_forward
[rank0]:     if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
[rank0]:                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]: RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by
[rank0]: making sure all `forward` function outputs participate in calculating loss.
[rank0]: If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
[rank0]: Parameter indices which did not receive grad for rank 0: 660 671 672 673 674 675 676 689 690 691 692 693 694 695 696 697 698 699 700 701 702 703 704 705 706 707 708 709 710 711 712 713 714 715 716 717 718 719 720 721 722 723 724
[rank0]:  In addition, you can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print out information about which particular parameters did not receive gradient on this rank as part of this error
W0627 23:22:48.127000 22841172398144 torch/distributed/elastic/multiprocessing/api.py:851] Sending process 2180803 closing signal SIGTERM
E0627 23:22:48.441000 22841172398144 torch/distributed/elastic/multiprocessing/api.py:826] failed (exitcode: 1) local_rank: 0 (pid: 2180802) of binary: /localscratch/aparek8.35464805.0/ultralyticsENV/bin/python
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/distributed/run.py", line 883, in <module>
    main()
  File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 347, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/distributed/run.py", line 879, in main
    run(args)
  File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/distributed/run.py", line 870, in run
    elastic_launch(
  File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/distributed/launcher/api.py", line 132, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/torch/distributed/launcher/api.py", line 263, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
/home/aparek8/.config/Ultralytics/DDP/_temp_jkws7orf22759235773008.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-06-27_23:22:48
  host      : cdr2539.int.cedar.computecanada.ca
  rank      : 0 (local_rank: 0)
  exitcode  : 1 (pid: 2180802)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
Traceback (most recent call last):
  File "/project/6000603/aparek8/ultralytics_CustomData/RTDETRx_RoadlabSignal_50.py", line 5, in <module>
    model.train(data="datasets/ROADLAB_TrafficSignalClassification/roadlabTSC.yaml", epochs=50, name="RTDETRx_RoadlabSignal_50", device=[0,1])
  File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/ultralytics/engine/model.py", line 650, in train
    self.trainer.train()
  File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/ultralytics/engine/trainer.py", line 199, in train
    raise e
  File "/localscratch/aparek8.35464805.0/ultralyticsENV/lib/python3.11/site-packages/ultralytics/engine/trainer.py", line 197, in train
    subprocess.run(cmd, check=True)
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2023/x86-64-v3/Compiler/gcccore/python/3.11.5/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/localscratch/aparek8.35464805.0/ultralyticsENV/bin/python', '-m', 'torch.distributed.run', '--nproc_per_node', '2', '--master_port', '47929', '/home/aparek8/.config/Ultralytics/DDP/_temp_jkws7orf22759235773008.py']' returned non-zero exit status 1.
```

### Environment

```
Ultralytics YOLOv8.2.46 🚀 Python-3.11.5 torch-2.3.0 CPU (Intel Xeon E5-2683 v4 2.10GHz)
Setup complete ✅ (64 CPUs, 251.8 GB RAM, 8.3/12.0 GB disk)

OS                  Linux-5.14.0-362.24.2.el9_3.x86_64-x86_64-Intel-R-_Xeon-R-_CPU_E5-2683_v4_@_2.10GHz-with-glibc2.37
Environment         Linux
Python              3.11.5
Install             pip
RAM                 251.78 GB
CPU                 Intel Xeon E5-2683 v4 2.10GHz
CUDA                None

numpy               ✅ 1.26.4+computecanada<2.0.0,>=1.23.0
matplotlib          ✅ 3.9.0+computecanada>=3.3.0
opencv-python       ✅ 4.9.0>=4.6.0
pillow              ✅ 10.3.0+computecanada>=7.1.2
pyyaml              ✅ 6.0.1+computecanada>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.13.1+computecanada>=1.4.1
torch               ✅ 2.3.0+computecanada>=1.8.0
torchvision         ✅ 0.18.0+computecanada>=0.9.0
tqdm                ✅ 4.66.4+computecanada>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0+computecanada
pandas              ✅ 2.2.1+computecanada>=1.1.4
seaborn             ✅ 0.13.2+computecanada>=0.11.0
ultralytics-thop    ✅ 2.0.0>=2.0.0
```

GPU info won't show up here since GPUs are allocated at runtime on my server.

### Minimal Reproducible Example

```
from ultralytics import RTDETR

model = RTDETR("models/rtdetr/rtdetr-x.pt")

model.train(data="datasets/ROADLAB_TrafficSignalClassification/roadlabTSC.yaml", epochs=50, name="RTDETRx_RoadlabSignal_50", device=[0,1])
```

### Additional

I found a similar issue that was closed before, and I replied in a comment, but I thought it would be better to raise a new issue.

https://github.com/ultralytics/ultralytics/issues/6296

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@AryanParekh hi there,

Thank you for providing detailed information about the issue you're facing with the RTDETR model. It seems like you've already taken several steps to troubleshoot, which is great! Let's try to address the problem further.

The error you're encountering, `RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one`, typically indicates that some parameters in your model are not being used in the loss computation. This can sometimes happen in multi-GPU setups with `torch.nn.parallel.DistributedDataParallel`.

Here are a few steps you can take to resolve this:

1. **Enable Unused Parameter Detection**: As suggested in the error message, you can enable unused parameter detection by passing `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`. This can help identify if there are any parameters not being used in the forward pass.
   
   ```python
   from ultralytics import RTDETR
   import torch

   model = RTDETR("models/rtdetr/rtdetr-x.pt")
   model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[0, 1], find_unused_parameters=True)

   model.train(data="datasets/ROADLAB_TrafficSignalClassification/roadlabTSC.yaml", epochs=50, name="RTDETRx_RoadlabSignal_50", device=[0,1])
   ```

2. **Verify Latest Versions**: Ensure that you are using the latest versions of all relevant packages. Sometimes, issues are resolved in newer releases. You mentioned using the latest version of Ultralytics, but please double-check that all dependencies are up to date.

3. **Minimal Reproducible Example**: You've provided a minimal reproducible example, which is very helpful. If the issue persists, please try running the example on a smaller subset of your data to see if the problem can be isolated further.

4. **Check Model Configuration**: Since you haven't made any internal changes to the model structure, it's worth verifying that the configuration files and parameters are correctly set up for the RTDETR model. Sometimes, minor discrepancies in configuration can lead to such issues.

5. **Consult Documentation**: For more detailed guidance on setting up multi-GPU training and handling such errors, you can refer to the [PyTorch DistributedDataParallel documentation](https://pytorch.org/docs/stable/notes/ddp.html).

If the issue persists after trying these steps, please let us know. Providing any additional logs or error messages can also help in diagnosing the problem further.
